test/pool_SUITE.erl

```diff
@@ -0,0 +1,311 @@
+%% Copyright (c) 2021, Loïc Hoguin <essen@ninenines.eu>
+%%
+%% Permission to use, copy, modify, and/or distribute this software for any
+%% purpose with or without fee is hereby granted, provided that the above
+%% copyright notice and this permission notice appear in all copies.
+%%
+%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+%% ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+%% OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+-module(pool_SUITE).
+-compile(export_all).
+-compile(nowarn_export_all).
+
+-import(ct_helper, [doc/1]).
+-import(ct_helper, [config/2]).
+-import(gun_test, [receive_from/1]).
+
+all() ->
+	ct_helper:all(?MODULE).
+
+init_per_suite(Config) ->
+	{ok, _} = cowboy:start_clear({?MODULE, tcp}, [], do_proto_opts()),
+	Port = ranch:get_port({?MODULE, tcp}),
+	[{port, Port}|Config].
+
+end_per_suite(_) ->
+	ExtraListeners = [
+		max_streams_h2_size_1,
+		max_streams_h2_size_2,
+		reconnect_h1
+	],
+	_ = [cowboy:stop_listener(Listener) || Listener <- ExtraListeners],
+	ok.
+
+do_proto_opts() ->
+	Routes = [
+		{"/", hello_h, []},
+		{"/delay", delayed_hello_h, 3000},
+		{"/ws", ws_echo_h, []}
+	],
+	#{
+		env => #{dispatch => cowboy_router:compile([{'_', Routes}])}
+	}.
+
+%% Tests.
+
+hello_pool_h1(Config) ->
+	doc("Confirm the pool can be used for HTTP/1.1 connections."),
+	Port = config(port, Config),
+	{ok, ManagerPid} = gun_pool:start_pool("localhost", Port, #{
+		protocols => [http],
+		scope => ?FUNCTION_NAME
+	}),
+	do_wait_pool_up(ManagerPid),
+	%% @todo Should pool functions return ok|error tuples?
+	Streams = [gun_pool:get("/",
+		#{<<"host">> => ["localhost:", integer_to_binary(Port)]},
+		#{scope => ?FUNCTION_NAME}
+	) || _ <- lists:seq(1, 8)],
+	_ = [begin
+		{response, nofin, 200, _} = gun_pool:await(StreamRef),
+		{ok, <<"Hello world!">>} = gun_pool:await_body(StreamRef)
+	end || StreamRef <- Streams].
+
+hello_pool_h2(Config) ->
+	doc("Confirm the pool can be used for HTTP/2 connections."),
+	Port = config(port, Config),
+	{ok, ManagerPid} = gun_pool:start_pool("localhost", Port, #{
+		protocols => [http2],
+		scope => ?FUNCTION_NAME
+	}),
+	do_wait_pool_up(ManagerPid),
+	%% @todo Should pool functions return ok|error tuples?
+	Streams = [gun_pool:get("/",
+		#{<<"host">> => ["localhost:", integer_to_binary(Port)]},
+		#{scope => ?FUNCTION_NAME}
+	) || _ <- lists:seq(1, 800)],
+	_ = [begin
+		{response, nofin, 200, _} = gun_pool:await(StreamRef),
+		{ok, <<"Hello world!">>} = gun_pool:await_body(StreamRef)
+	end || StreamRef <- Streams].
+
+hello_pool_ws(Config) ->
```

Comment:
You can also use it with Websocket but it requires the use of a Websocket handler (above) to forward messages to the correct process.